src/lib/adopted-probes.ts

```diff
@@ -1,49 +1,122 @@
 import type { Knex } from 'knex';
 import Bluebird from 'bluebird';
 import _ from 'lodash';
-
 import { scopedLogger } from './logger.js';
 import { client } from './sql/client.js';
-import { fetchSockets } from './ws/server.js';
+import { fetchRawSockets } from './ws/server.js';
 import type { Probe } from '../probe/types.js';
+import { normalizePublicName } from './geoip/utils.js';
 
 const logger = scopedLogger('adopted-probes');
 
-const TABLE_NAME = 'adopted_probes';
+export const ADOPTED_PROBES_TABLE = 'adopted_probes';
+export const USERS_TABLE = 'directus_users';
 
-type AdoptedProbe = {
+export type AdoptedProbe = {
+	username: string;
 	ip: string;
 	uuid: string;
 	lastSyncDate: string;
+	tags: string[];
+	isCustomCity: boolean;
 	status: string;
 	version: string;
 	country: string;
-	city: string;
-	latitude: number;
-	longitude: number;
+	city?: string;
+	state?: string;
+	latitude?: number;
+	longitude?: number;
 	asn: number;
 	network: string;
 }
 
+type Row = Omit<AdoptedProbe, 'isCustomCity' | 'tags'> & {
+	tags: string;
+	isCustomCity: number;
+}
+
 export class AdoptedProbes {
 	private connectedIpToProbe: Map<string, Probe> = new Map();
 	private connectedUuidToIp: Map<string, string> = new Map();
+	private adoptedIpToProbe: Map<string, AdoptedProbe> = new Map();
 	private readonly adoptedFieldToConnectedField = {
-		status: 'status',
-		version: 'version',
-		country: 'location.country',
-		city: 'location.city',
-		latitude: 'location.latitude',
-		longitude: 'location.longitude',
-		asn: 'location.asn',
-		network: 'location.network',
+		status: {
+			connectedField: 'status',
+			shouldUpdateIfCustomCity: true,
+		},
+		version: {
+			connectedField: 'version',
+			shouldUpdateIfCustomCity: true,
+		},
+		asn: {
+			connectedField: 'location.asn',
+			shouldUpdateIfCustomCity: true,
+		},
+		network: {
+			connectedField: 'location.network',
+			shouldUpdateIfCustomCity: true,
+		},
+		country: {
+			connectedField: 'location.country',
+			shouldUpdateIfCustomCity: false,
+		},
```

Comment:
If countries of our geo resolution and adoption are different - adoption data is not applied. So result will be C2 with the city from geo resolution logic.
At the same time adopted probe city/country data is not changed (just not used). So if country goes back to C1 city will be A2 again.

Reply:
I think we should update the country in the DB to C2 (but keep the city as set by the user) so that the user can readjust the city in C2 if needed.

Reply:
Is that ok to display city not from the displayed country on the dashboard?

Plus, what is the use case for that? User changes the city to the correct A2 in C1. But when geo logic says it is another country C2, user founds another correct city?

Reply:
> Is that ok to display city not from the displayed country on the dashboard?

I think that's exactly what we need. If the dashboard says C1, but the probe actually can't be targeted via C1 because API sees C2, it'll look broken and confusing. The dashboard should always show the current state.

> user founds another correct city?

Could be the IP got moved... But I agree that's very rare. More likely is that C2 is wrong, and in that case the previous point is important - it should be visible in the dashboard otherwise it'll be confusing.

Reply:
> If the dashboard says C1, but the probe actually can't be targeted via C1 because API sees C2, it'll look broken and confusing

Same can be applied to cities. 

We can add something like `editIsActive` flag to indicate status of custom city, but that may confuse even more.

I personally think not applying change if country is different without updates on dashboard is the simples logic to understand.

Reply:
>Same can be applied to cities.

Yes, it would be best to show a warning in that case as well. Otherwise, it won't be clear to anyone what happened when it doesn't work.

Reply:
So we are only sending notification (similar to gh username change notification) and ignoring city and country from the adoption? Or additionally changing the country on the dashboard?

Reply:
Let's change country in dashboard, notify and link to https://github.com/jsdelivr/globalping/issues/268 in the notification (If this change is not right, you can report it at ...).